buildpacks/java/tests/detect_test.go

```diff
@@ -0,0 +1,201 @@
+	appDir, cleanup := SetupTestDirectory(
+		WithFuncName("my-function"),
+	)
+	defer cleanup()
+
+	plan, err := d.Detect(getContext(
+		withApplicationPath(appDir),
+	))
+
+	expectations := DetectExpectations{
+		Result: result{plan, err},
+		Pass:   true,
+		Metadata: map[string]interface{}{
+			"func_yaml_function_name": "my-function",
+			"func_yaml_envs":          map[string]string{},
+			"func_yaml_options":       map[string]string{},
+		},
+	}
+	expectations.Check(t)
+}
+
+func TestDetectEnvironmentWithValidFile(t *testing.T) {
+	d := getDetector()
+	appDir, cleanup := SetupTestDirectory(
+		WithFuncEnvs(map[string]string{
+			"MODULE_NAME":   "other",
+			"FUNCTION_NAME": "handler2",
+		}),
+	)
+	defer cleanup()
+
+	plan, err := d.Detect(getContext(
+		withApplicationPath(appDir),
+		withModuleName("other"),
+		withFunctionName("handler2"),
+	))
+
+	expectations := DetectExpectations{
+		Result: result{plan, err},
+		Pass:   true,
+		Metadata: map[string]interface{}{
+			"func_yaml_function_name": "handler2",
+			"func_yaml_envs": map[string]string{
+				"MODULE_NAME":   "other",
+				"FUNCTION_NAME": "handler2",
+			},
+			"func_yaml_options": map[string]string{},
+		},
+	}
+	expectations.Check(t)
+}
+
+func TestDetectNameFromEnvsBeforeFuncYamlFile(t *testing.T) {
+	d := getDetector()
+	appDir, cleanup := SetupTestDirectory(
+		WithFuncName("not-the-best-name"),
+		WithFuncEnvs(map[string]string{
+			"MODULE_NAME":   "other",
+			"FUNCTION_NAME": "goodName",
+		}),
+	)
+	defer cleanup()
+
+	plan, err := d.Detect(getContext(
+		withApplicationPath(appDir),
+		withModuleName("other"),
+		withFunctionName("goodName"),
+	))
+
+	expectations := DetectExpectations{
+		Result: result{plan, err},
+		Pass:   true,
+		Metadata: map[string]interface{}{
+			"func_yaml_function_name": "goodName",
+			"func_yaml_envs": map[string]string{
+				"MODULE_NAME":   "other",
+				"FUNCTION_NAME": "goodName",
+			},
+			"func_yaml_options": map[string]string{},
+		},
+	}
+	expectations.Check(t)
+}
+
+func getDetector() java.Detect {
+	buf := bytes.NewBuffer(nil)
+	logger := bard.NewLogger(buf)
+	return java.Detect{
+		Logger: logger,
+	}
+}
+
+func getContext(opts ...func(*libcnb.DetectContext)) libcnb.DetectContext {
+	ctx := libcnb.DetectContext{
+		Application: libcnb.Application{},
+		Buildpack:   libcnb.Buildpack{},
+		Platform: libcnb.Platform{
+			Environment: make(map[string]string),
+		},
+	}
+
+	for _, opt := range opts {
+		opt(&ctx)
+	}
+
+	return ctx
+}
+
+func withApplicationPath(path string) func(*libcnb.DetectContext) {
+	return func(dc *libcnb.DetectContext) {
+		dc.Application.Path = path
+	}
+}
+
+func withModuleName(value string) func(*libcnb.DetectContext) {
+	return func(dc *libcnb.DetectContext) {
+		dc.Platform.Environment["MODULE_NAME"] = value
+	}
+}
+
+func withFunctionName(value string) func(*libcnb.DetectContext) {
+	return func(dc *libcnb.DetectContext) {
+		dc.Platform.Environment["FUNCTION_NAME"] = value
+	}
+}
```

Comment:
These are only relevant to Python

Reply:
Oh, the only way in `Java` to set the function name is with `BP_FUNCTION` or the Spring property?